FungalTree.wdl
=========================
The FungalTree workflow performs variant calling and phylogenetic tree construction from paired-end reads for fungal haploid genomes. It aligns reads to a reference genome, performs variant calling using GATK HaplotypeCaller, applies hard filters, and generates a multiple sequence alignment (MSA) for phylogenetic inference using IQ-TREE 2.


## Task Descriptions and Enhancements

### Task: [GbffToFasta]()

**Purpose:**
Convert a GenBank-format reference file to a FASTA format for downstream processing.

**Inputs:**
- `ref_gbff`: Reference file in GenBank (.gbff) format.

**Outputs:**
- `reference_fasta`: Reference genome in FASTA format.


### Task: [GenerateRefFiles]()

**Purpose:**
This task prepares all required reference genome index files needed for downstream tools including BWA, Samtools, and Picard.

**Inputs:**
- `ref_fasta`: Reference genome in FASTA format
- `input_bam`: Aligned bam file

**Outputs:**
- `ref_sa`: BWA suffix array index (.sa)
- `ref_bwt`: BWA Burrows-Wheeler Transform (.bwt)
- `ref_amb`: BWA ambiguity file (.amb)
- `ref_ann`: BWA annotation file (.ann)
- `ref_pac`: BWA packed sequence (.pac)
- `ref_dict`: Sequence dictionary (.dict) generated by Picard
- `ref_index`: FASTA index file (.fai) generated by Samtools
- `reference_fasta`: The input reference genome


