FungalTree.wdl
=========================
The FungalTree workflow performs variant calling and phylogenetic tree construction from paired-end reads for fungal haploid genomes. It aligns reads to a reference genome, performs variant calling using GATK HaplotypeCaller, applies hard filters, and generates a multiple sequence alignment (MSA) for phylogenetic inference using IQ-TREE 2.


## Task Descriptions and Enhancements

### Task: [GbffToFasta]()

**Purpose:**
Convert a GenBank-format reference file to a FASTA format for downstream processing.

**Inputs:**
- `ref_gbff`: Reference file in GenBank (.gbff) format.

**Outputs:**
- `reference_fasta`: Reference genome in FASTA format.


### Task: [GenerateRefFiles]()

**Purpose:**
This task prepares all required reference genome index files needed for downstream tools including BWA, Samtools, and Picard.

**Inputs:**
- `ref_fasta`: Reference genome in FASTA format
- `input_bam`: Aligned bam file

**Outputs:**
- `ref_sa`: BWA suffix array index (.sa)
- `ref_bwt`: BWA Burrows-Wheeler Transform (.bwt)
- `ref_amb`: BWA ambiguity file (.amb)
- `ref_ann`: BWA annotation file (.ann)
- `ref_pac`: BWA packed sequence (.pac)
- `ref_dict`: Sequence dictionary (.dict) generated by Picard
- `ref_index`: FASTA index file (.fai) generated by Samtools
- `reference_fasta`: The input reference genome


### Task: [VCFToFasta]()

**Purpose:**
This task converts a multi-sample VCF file into a multiple sequence alignment (MSA) FASTA format suitable for phylogenetic analysis.

**Inputs:**
- `vcf_file`: A gzipped multi-sample VCF file (.vcf.gz)

**Outputs:**
- `alignmnent_fasta`: MSA FASTA file containing genotype-based sequences per sample. The .min4.fasta suffix indicates a minimum of 4 alleles per site retained (as per the scriptâ€™s default filtering behavior).



### Task: [IqTree2]()

**Purpose:**
Constructs a robust phylogenetic tree from aligned variant sequences with statistical support values, forming the final output of the variant-to-tree workflow.

**Inputs:**
- `alignment`: FASTA file with aligned sequences (e.g. output from VCFToFasta)
- `cluster_name`: Prefix for output tree filename
- `iqtree2_model`:  Optional string. Substitution model to use. If not provided, IQ-TREE's ModelFinder will determine the best-fit model.
- `iqtree2_bootstraps`: Number of ultrafast bootstrap replicates (default = 1000)
- `alrt`: Number of SH-like approximate likelihood ratio test replicates (default = 1000)
- `iqtree2_opts`: Optional String. User-supplied options for IQ-TREE

**Outputs:**
- `ml_tree`: Newick file containing the final phylogenetic tree (*_iqtree.nwk)
- `iqtree2_model_used`: Substitution model used (either user-specified or inferred by ModelFinder)
- `iqtree2_version`: IQ-TREE 2 version string
- `date`: Timestamp of the run
